vn.hoabinh.quan.CoBang.yaml

```diff
@@ -0,0 +1,90 @@
+app-id: vn.hoabinh.quan.CoBang
+runtime: org.gnome.Platform
+runtime-version: '3.36'
+sdk: org.gnome.Sdk
+command: cobang
+finish-args:
+  - --share=network
+  - --share=ipc
+  - --socket=wayland
+  - --socket=fallback-x11
+  # Need to load image files
+  - --filesystem=home:ro
```

Comment:
You shouldn't need this to load image files using GTK. If you switch to using https://developer.gnome.org/gtk3/stable/gtk3-GtkFileChooserNative.html then the portal does all of this for you automatically and you can close this hole.

Reply:
Is `Gtk.FileChooserNative` the only allowed approach in Flatpak? If not, please accept the current implementation. It is because the Gtk dialog is from `Gtk.FileChooseButton`, which I'm using.

I try to make `Gtk.FileChooserButton` use `GtkFileChooserNative` but it refuses:

```python
  File "/home/quan/Works/Hobby/CoBang/CoBang/cobang/app.py", line 177, in build_filechooser
    button: Gtk.FileChooserButton = Gtk.FileChooserButton.new_with_dialog(dialog)
TypeError: argument dialog: Expected Gtk.Dialog, but got gi.repository.Gtk.FileChooserNative
```

Reply:
I think that I just disable this feature for Flatpak-installed version. I've just tested with `flatpak-builder --run` and found that, even with `--filesystem=home:ro`, the original `FileChooser` cannot retrieve the image:

```py
Traceback (most recent call last):
  File "/home/quan/Works/Hobby/CoBang/CoBang/cobang/app.py", line 401, in on_btn_img_chooser_file_set
    self.process_passed_image(chosen_file)
  File "/home/quan/Works/Hobby/CoBang/CoBang/cobang/app.py", line 373, in process_passed_image
    stream: Gio.FileInputStream = chosen_file.read(None)
gi.repository.GLib.Error: g-io-error-quark: Error opening file /run/user/1000/doc/8d26bd81/25-3-quet-qr-code-de-tai-ung-dung-ncovi.jpg: No such file or directory (1)
Traceback (most recent call last):
  File "/home/quan/Works/Hobby/CoBang/CoBang/cobang/app.py", line 401, in on_btn_img_chooser_file_set
    self.process_passed_image(chosen_file)
  File "/home/quan/Works/Hobby/CoBang/CoBang/cobang/app.py", line 373, in process_passed_image
    stream: Gio.FileInputStream = chosen_file.read(None)
gi.repository.GLib.Error: g-io-error-quark: Error opening file /run/user/1000/doc/bad8a6f6/Screenshot_2020-05-19-23-44-49-880_com.android.settings.jpg: No such file or directory (1)
```

Will recommend user to install from traditional way to get full feature.

Reply:
`flatpak-builder --run` is not accurate way to test permissions.

Reply:
How are you trying to open that file? per the log you seem to have `gi.repository.GLib.Error: g-io-error-quark: Error opening file /run/user/1000/doc/bad8a6f6/Screenshot_2020-05-19-23-44-49-880_com.android.settings.jpg: No such file or directory (1)` which shouldn't happen. If you're using the NativeFileChooser, the portal grants you access to the file and store that on the permissions store.

If you use the `get_file` method of NativeFileChooser or you create a `gio.File` yourself using the uri of the file it should work. You should not the file path as it's not something you can't access

Reply:
It's not mandatory to use the portal currently, just provides a nicer experience. I'm surprised that doesn't work for you though.

Reply:
Yes, I use `Gtk.FileChooser.get_file` to get a `Gio.File`, like in this signal handler

https://github.com/hongquan/CoBang/blob/afffa4d5e5873c71f03eb2518a5a9b8f77ef437f/cobang/app.py#L398

```py
def on_btn_img_chooser_file_set(self, chooser: Gtk.FileChooserButton):
	chosen_file: Gio.File = chooser.get_file()
	logger.debug('Chose file: {}', chosen_file.get_uri())
	self.process_passed_image(chosen_file)
```

The problem is that, for non-Flatpak environment, the `file.get_uri()` return exactly a URI, like `file://` or `sftp://`, but with Flatpak environment, it return a file path.

May be it is because I'm using `flatpak-builder --run` to test? Which is the right command?

Reply:
Both are a URI, there must be an issue somewhere but it's not really a blocker. You can look at that for a future release